app/Main.hs

```diff
@@ -2,12 +2,58 @@ import qualified Data.ByteString.Lazy as BSL
 import qualified Data.ByteString.Lazy.Char8 as BSL8
 
 import Control.Concurrent
-import System.Environment
+import Options.Applicative
 import System.FilePath
 
 import Network.SSH.Client.LibSSH2.Foreign
 import Network.SSH.Client.LibSSH2
 
+-- Data types for CLI options
+
+data Options = Options {
+  connectionInfo :: Connection,
+  keys           :: KeyFiles,
+  script         :: FilePath,
+  logFile        :: FilePath
+} deriving (Show)
+
+data Connection = Connection {
+  user :: String,
+  host :: String,
+  port     :: Int
+} deriving (Show)
+
+data KeyFiles = KeyFiles {
+  publicKey  :: FilePath,
+  privateKey :: FilePath
+} deriving (Show)
+
+-- Parsers for CLI options
+
+connectionParser :: Parser Connection
+connectionParser = Connection <$> userParser <*> hostParser <*> portParser
+  where
+    userParser = strOption (long "user" <> help "Username")
```

Comment:
As well as the associative operator here 💯

Reply:
The `opt-applicative` library is the star here - I'm just following their recommended patterns. 😄